cpp/src/join/join.cu

```diff
@@ -26,15 +26,47 @@
 namespace cudf {
 namespace detail {
 
-std::unique_ptr<table> inner_join(
-  table_view const& left_input,
-  table_view const& right_input,
-  std::vector<size_type> const& left_on,
-  std::vector<size_type> const& right_on,
-  std::vector<std::pair<size_type, size_type>> const& columns_in_common,
-  null_equality compare_nulls,
-  rmm::cuda_stream_view stream,
-  rmm::mr::device_memory_resource* mr)
+std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+          std::unique_ptr<rmm::device_uvector<size_type>>>
+inner_join(table_view const& left_input,
+           table_view const& right_input,
+           null_equality compare_nulls,
+           rmm::cuda_stream_view stream,
+           rmm::mr::device_memory_resource* mr)
+{
+  // Make sure any dictionary columns have matched key sets.
+  // This will return any new dictionary columns created as well as updated table_views.
+  auto matched = cudf::dictionary::detail::match_dictionaries(
+    {left_input, right_input},
+    stream,
+    rmm::mr::get_current_device_resource());  // temporary objects returned
+
+  // now rebuild the table views with the updated ones
+  auto const left  = matched.second.front();
+  auto const right = matched.second.back();
+
+  // For `inner_join`, we can freely choose either the `left` or `right` table to use for
+  // building/probing the hash map. Because building is typically more expensive than probing, we
+  // build the hash map from the smaller table.
+  if (right.num_rows() > left.num_rows()) {
+    cudf::hash_join hj_obj(left, compare_nulls, stream);
+    auto result = hj_obj.inner_join(right, compare_nulls, stream, mr);
+    return std::make_pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+                          std::unique_ptr<rmm::device_uvector<size_type>>>(std::move(result.second),
+                                                                           std::move(result.first));
```

Comment:
The explicit template args should not be necessary here.

Reply:
Fixed